### ISSUE TITLE
Conditionally close the retry and circuit breaker policy in 4.3.0 version for 3.5 framework

### ISSUE THREAD

**Issue by soundarpandiyan** [App-vNext/Polly#626]
**Summary: What are you wanting to achieve?**
Conditionally close the waitandretryforever and circuit breaker policy  in 4.3.0 version for 3.5 framework
<br/>
<br/>

**What code or approach do you have so far?**  
```
 circuitBreakerPolicy = Policy
               .Handle<Exception>()
               .CircuitBreaker(
                   exceptionsAllowedBeforeBreaking: 2,
                   durationOfBreak: TimeSpan.FromSeconds(3),
                   onBreak: (ex, breakDelay, context) =>
                   {
                       //offline
                       Console.WriteLine(" offline Breaker logging: Breaking the circuit for " + breakDelay.TotalMilliseconds + "ms!" + "due to" + ex.Message);
                   },
                   onReset: (context) => HandleResetState(),
                   onHalfOpen: () => HandleHalfOpenState()
               );

              policy = Policy.Handle<WebException>()
                               .WaitAndRetryForever(
                                    retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
                                    (exception, timeSpan, context) =>
                                    {
                                        Console.WriteLine("Encountered exception {0}", exception.Message);
                                    });

policy.Execute(() => circuitBreakerPolicy.Execute(() => ExecuteHTTPGetWithFlowThroughException(messages[0])));
```
<br/>
<br/>

**Comment by reisenberger** [comment]
@soundarpandiyan Many apologies: I have not understood the question.  Could you explain, with more detail, what you are wanting to achieve? (That is: Expand more about "Conditionally close the waitandretryforever and circuit breaker"). Thank you.

**Comment by soundarpandiyan** [comment]
Apologies for shortwritten. provided an example below. 

Assume it is a shopping cart service. I would like to hit an API call to get user profile information once user logged in. 

User logged in
Hitting a /userinfo API to get profile information using RetryForever and CircuitBreaker policy.
Server is down so its keep on retrying 
User performs logged out so i need to cancel this existing http call which i triggered earlier

Query1:
Whether it is possible to cancel the http call based on log out condition in 4.3.0.

Query2:
whether the execution of policies added for that particular http request will get ended once the request become success? I have a scenario to keep on checking whether the server is online so we are trying a separate api to check for that server is alive or not. Do i need to add separate thread with time interval to keep on hitting the keep alive message. Does this possible in polly.

what im expecting is the polly execution should happen for predefined interval even if the message getting succeeded.

**Comment by reisenberger** [comment]
Thanks @soundarpandiyan for the explanation.  Re:

>Hitting a /userinfo API to get profile information using RetryForever and CircuitBreaker policy.
>User performs logged out so i need to cancel this existing http call which i triggered earlier
>Whether it is possible to cancel the http call based on log out condition in 4.3.0.

You can specify a condition for whether to continue retrying when an exception occurs, as shown in the [second example in step 1 of the readme](https://github.com/App-vNext/Polly#step-1--specify-the--exceptionsfaults-you-want-the-policy-to-handle).  For example:

    policy = Policy.Handle<WebException>(ex => IsLoggedOn(userId)).WaitAndRetryForever(...)

assuming some method `bool IsLoggedOn(userId)`.  This prevents any further retries when a `WebException` is encountered but the call to `IsLoggedOn(...)` returns `false`.  When that situation occurs, the `WebException` will be rethrown on to the calling code.  But it does not cancel any http request which is already in flight. That is not possible in Polly v4.3.0 for .NET3.5.  The closest functionality is [Timeout policy](https://github.com/App-vNext/Polly/wiki/Timeout) in [sychronous pessimistic mode](https://github.com/App-vNext/Polly/wiki/Timeout#pessimistic-timeout) which can be used to 'walk away from' waiting for (not cancel) an otherwise-uncancellable synchronous http call. This is only available in Polly v5.0+, for .NET Framework 4.5+.

Re:

>Query2:
>whether the execution of policies added for that particular http request will get ended once the request become success? 

Yes, it will end.  The behaviour is retry-until-succeeds.

>Do i need to add separate thread with time interval to keep on hitting the keep alive message? Does this possible in polly?
>what im expecting is the polly execution should happen for predefined interval even if the message getting succeeded.

Polly intentionally does not provide this facility. You can start a separate thread and/or use a `Timer(...)` to invoke work periodically. We have intentionally restricted the scope of Polly to managing single executions, not scheduling executions. We decided not to move Polly in the direction of scheduling executions because there are already a number of satisfactory offerings/solutions in .NET for persistent cron-job execution - although I don't know how many will be suitable for .NET 3.5. I was thinking of things like [Hangfire](https://github.com/HangfireIO/Hangfire) or [Quartz.Net](https://www.quartz-scheduler.net/) for .NET framework, if you need persistent storage behind the cron-jobs; or [`IHostedService` in dot net core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-2.2); and any number of timer-triggered offerings in Azure.

Let us know if that answers your question or if we can help with anything else.

**Comment by soundarpandiyan** [comment]
Thanks @reisenberger for the detail explanation on this.
